### ISSUE TITLE
Geeqie exits when overwriting an image file being displayed

### ISSUE THREAD

**Issue by jefferysmall** [BestImageViewer/geeqie#1221]
### ISSUE TYPE

* Bug Report

### GEEQIE VERSION

```
99-> geeqie --version
Gtk-Message: 18:07:38.913: Failed to load module "xapp-gtk3-module"
Geeqie 2.1+git20231221-214be330 GTK3

This is the extracted current AppImage
```
### OS / DISTRIBUTION
Xubuntu 22.04.3 LTS

### SUMMARY
Geeqie is open and displaying one of a number of images in a directory -- a combination of Sony raw (arw) and jpg files.  GIMP is being used to edit the jpg file being displayed.  When the jpg is exported from GIMP and overwrites the existing file, sometimes this happens gracefully and then other times it causes Geeqie to immediately exit.  There is no crash file produced.

I wonder if this is in some way related to bug [#159](https://github.com/BestImageViewer/geeqie/issues/159
)

### STEPS TO REPRODUCE
Export files from GIMP until it occurs.

**Comment by caclark** [comment]
I suspect this happens when `Edit/Preferences/General/Refresh On File Change` is selected.
This could be turned off, and a manual Refresh made at intervals.
I know this does not solve the problem.

If this is the case, perhaps the `realtime_monitor_cb()` is seeing the change and unreffing the `fd`, while the same `fd` remains in` vf->list`.
It might be the solution is to eliminate the `assert()` that causes the crash, and just ignore the error.

**Comment by jefferysmall** [comment]
I tried repeatedly saving an image that was not currently being displayed in Geeqie and no problems.

I tried your suggestion of disabling 'Refresh On File Change' and there were no exits/crashes.  However, If I saved and quickly clicked on the refresh button while the save was still underway, then Geeqie crashed.  The need for a manual refresh is a pain and I will revert to the previous version until this is addressed.

I just upgraded to this development version two days ago in order to (hopefully) address some other problems, and began to see this crashing problem.  For reference,, this never occurred with version 1.7+git20220327-6d541232 GTK3

Thanks for looking into this.

**Comment by caclark** [comment]
I am still unable to replicate this.
The latest AppImages have the -Ddevel=enabled option set, which produces a (usually useful) stack dump when a crash occurs.

**Comment by jefferysmall** [comment]
OK, I'll download the most current AppImage and see what happens.  If it does core dump will it be in the starting directory or elsewhere, and how should I get it to you?

**Comment by jefferysmall** [comment]
Tried it running 2.2+git20240207-5547f6ae (extracted AppImage) and it exited on the first attempt to save an image.  Retried and same thing happened.

However, there is no dump file in the current directory, in /var/lib/apport/coredump or in /var/crash.

**Comment by jefferysmall** [comment]
Now using Geeqie 2.4+git20240624-f47c8f1a GTK3 (extracted AppImage) and I'm getting the same issue.  When the thumbnail is active and the current JPG image is exported repeatedly from GIMP, geeqie core dumps.  I got a core image in /var/lib/apport/coredump.  here is some gdb info:

% file /usr/lbin/Geeqie-240624/AppRun
/usr/lbin/Geeqie-240624/AppRun: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=86bc6644a3eca0daef30e3b0ec672a7eb121044f, not stripped

% gdb  /usr/lbin/Geeqie-240624/AppRun  /var/lib/apport/coredump/core*
...
Reading symbols from /usr/lbin/Geeqie-240624/AppRun...
(No debugging symbols found in /usr/lbin/Geeqie-240624/AppRun)

warning: Can't open file /SYSV00000000 (deleted) during file-backed mapping note processing

warning: core file may not match specified executable file.
[New LWP 3869531]
[New LWP 3869074]
[New LWP 3869060]
[New LWP 3869054]
Core was generated by `/usr/local/bin/Geeqie-240624/usr/local/bin/geeqie --geometry=1852x2039+989+41 -'.
Program terminated with signal SIGBUS, Bus error.
#0  0x00007fb4132089fc in ?? ()
(gdb) bt
#0  0x00007fb4132089fc in ?? ()
#1  0x0000000000000000 in ?? ()

Not very useful.  Suggestions?  I can upload the core file if it would be helpful.

**Comment by jefferysmall** [comment]
I tried:

% /usr/lbin/Geeqie-240624/AppRun  -Ddevel=enabled  .

But I get a pop-up that says:

Invalid parameter(s):
-Ddevel=enabled
This option is unknown

**Comment by xsdg** [comment]
Howdy.  According to <https://askubuntu.com/a/1231605>, you can run an AppImage with `--appimage-extract`, which will hopefully extract a `geeqie` binary somewhere in there.  Try using that binary with gdb and your core file.

That said, sometimes a Bus Error is a hardware issue of some sort.

**Comment by jefferysmall** [comment]
xsdg:  That is exactly what I did and reported two comments ago.  The AppRun is the extracted binary and the gdb output is what I got against the core file.  It looks like this version of the AppImage may not have been compiled with the proper settings.  And even thought the binary is not stripped, gdb cannot find any debugging symbols.

**Comment by xsdg** [comment]
Are you positive?  Your gdb output shows:
```Core was generated by `/usr/local/bin/Geeqie-240624/usr/local/bin/geeqie --geometry=1852x2039+989+41 -'.```

So I was expecting the geeqie binary to be either at that relative location, or in some other subdirectory.  (That said, note that I have no experience with AppImages, so anything I say/infer could easily be wrong)

**Comment by jefferysmall** [comment]
You are correct.  I assumed that /usr/lbin/Geeqie-240624//AppRun was the actual binary but it must exec /usr/lbin/Geeqie-240624/usr/local/bin/geeqie.  I should have looked more closely at the preliminary info.  Let's try again:

% gdb /usr/lbin/Geeqie-240624/usr/local/bin/geeqie  /var/lib/apport/coredump/core._usr_local_bin_Geeqie-240624_usr_local_bin_geeqie.1000.7eba8768-e63b-41a1-89bc-debc2c3c5730.3869054.34085054

GNU gdb (Ubuntu 12.1-0ubuntu1~22.04.2) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/lbin/Geeqie-240624/usr/local/bin/geeqie...

Warning: Can't open file /SYSV00000000 (deleted) during file-backed mapping note processing
[New LWP 3869531]
[New LWP 3869074]
[New LWP 3869060]
[New LWP 3869054]

warning: Could not load shared library symbols for lib64/ld-linux-x86-64.so.2.
Do you need "set solib-search-path" or "set sysroot"?

warning: Loadable section ".note.gnu.build-id" outside of ELF segments
  in /usr/local/bin/Geeqie-240624/runtime/compat/lib/x86_64-linux-gnu/libc.so.6
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/local/bin/Geeqie-240624/usr/local/bin/geeqie --geometry=1852x2039+989+41 -'.
Program terminated with signal SIGBUS, Bus error.
#0  __pthread_kill_implementation (no_tid=0, signo=7, threadid=140411106801216)
    at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
[Current thread is 1 (Thread 0x7fb40224b640 (LWP 3869531))]

(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=7, threadid=140411106801216)
    at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=7, threadid=140411106801216)
    at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=140411106801216, signo=signo@entry=7)
    at ./nptl/pthread_kill.c:89
#3  0x00007fb4131b4476 in __GI_raise (sig=7) at ../sysdeps/posix/raise.c:26
#4  0x000055bb62e1e4ba in add_collection_list (
    collection_list=<optimized out>, func=0x7fb402249910, menu=0x4)
    at ../src/menu.cc:438
#5  submenu_add_collections (menu=<optimized out>, menu_item=0x0, 
    func=0x7fb402249910, data=<optimized out>) at ../src/menu.cc:466
#6  <signal handler called>
#7  0x00007fb413debbcf in jpeg_fill_bit_buffer ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8
#8  0x00007fb413ded97c in ?? ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8
#9  0x00007fb413dde076 in ?? ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8
#10 0x00007fb413de62c6 in jpeg_start_decompress ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8
#11 0x000055bb62e84f22 in bar_pane_gps_view_state_changed_cb (
    view=<optimized out>, data=0x7fb402249c00) at ../src/bar-gps.cc:778
#12 0x000055bb62dee6a4 in image_loader_get_error (il=0x7fb402249910)
    at ../src/image-load.cc:1393
#13 0x000055bb62deef30 in (anonymous namespace)::ddsDecodeDXT3 (
    width=320890832, height=1686872112, 
    buffer=0x7ffc7230bfe0 "@\203\212d\273U") at ../src/image-load-dds.cc:281
#14 0x00007fb41475b714 in ?? ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#15 0x00007fb414758ab1 in ?? ()
   from /usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#16 0x00007fb413206ac3 in start_thread (arg=<optimized out>)
    at ./nptl/pthread_create.c:442
#17 0x00007fb413298850 in clone3 ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81

(gdb) q
%

**Comment by xsdg** [comment]
@jefferysmall What file formats are present in the directory that Geeqie has open?  (If you're willing to share a filename listing, that would be helpful, but it shouldn't be necessary).  Also, which file type (or filename) did Geeqie have open when the crash happened?

It looks like frame 15 is already on the wrong track (`width` and `height` aren't believable).  After that, I wouldn't be surprised if we're in Undefined Behavior land (perhaps with a corrupt stack).  In particular, we started off in the DDS image loader, and then bounced through bar-gps.cc to end up in libjpeg, even though the method in question (`bar_pane_gps_view_state_changed_cb`) doesn't seem to have any method calls other than gtk UI state updates.  After that, we bounced through libjpeg to end up `submenu_add_collections` somehow, which also doesn't seem to make sense.

**Comment by jefferysmall** [comment]
The problem can be exhibited in any directory including one with a single JPG file.  My directories usually include a combination of JPG and ARW (Sony Raw) files.  Geeqie never has any problem opening up and displaying whatever files (including PDFs) are in the directory or displaying the selected image.  It only has this problem when the files is being exported from GIMP.  I tried placing a single file in a directory and then repeatedly and quickly touching it.  Geeqie did not seem to have a problem with this.  One thing you might investigate is that whne GIMP exports the file, it may be deleting or zeroing out the existing file when overwriting it and geeqie may be getting "lost" as the status of the file and thumbnail are updated.  Here is the listing from the directory that I was most recently working in.  I happened to be saving "bronners_15.jpg" over and over as my test when geeqie crashes.  I'm pretty sure this is a thumbnail problem since geeqie will not crash if any other thumbnail is selected while the export operation occurs.  A directory listing and a screenshot of geeqie follow.

116-> ls -l
total 1334028
-rw-r--r-- 1 jeff cjs 38704664 Oct 21  2023 bronners_01.arw
-rw-r--r-- 1 jeff cjs 16732001 Jun 27 15:03 bronners_01.jpg
-rw-r--r-- 1 jeff cjs 39391794 Oct 21  2023 bronners_02.arw
-rw-r--r-- 1 jeff cjs 15512296 Jun 27 15:13 bronners_02.jpg
-rw-r--r-- 1 jeff cjs 38410668 Oct 21  2023 bronners_03.arw
-rw-r--r-- 1 jeff cjs 15227024 Jun 28 17:00 bronners_03.jpg
-rw-r--r-- 1 jeff cjs 39121478 Oct 21  2023 bronners_04.arw
-rw-r--r-- 1 jeff cjs 17869581 Jun 28 17:03 bronners_04.jpg
-rw-r--r-- 1 jeff cjs 38140084 Oct 21  2023 bronners_05.arw
-rw-r--r-- 1 jeff cjs 14375359 Jun 28 17:07 bronners_05.jpg
-rw-r--r-- 1 jeff cjs 37419710 Jun 29 18:01 bronners_06.arw
-rw-r--r-- 1 jeff cjs 11673451 Jun 29 18:01 bronners_06.jpg
-rw-r--r-- 1 jeff cjs 38592578 Oct 21  2023 bronners_07.arw
-rw-r--r-- 1 jeff cjs 14484413 Jun 29 17:55 bronners_07.jpg
-rw-r--r-- 1 jeff cjs 39947182 Oct 21  2023 bronners_08.arw
-rw-r--r-- 1 jeff cjs 18821506 Jun 29 18:01 bronners_08.jpg
-rw-r--r-- 1 jeff cjs 39459374 Oct 21  2023 bronners_09.arw
-rw-r--r-- 1 jeff cjs 17366964 Jun 29 18:06 bronners_09.jpg
-rw-r--r-- 1 jeff cjs 39117896 Oct 21  2023 bronners_10.arw
-rw-r--r-- 1 jeff cjs 17676717 Jun 29 18:10 bronners_10.jpg
-rw-r--r-- 1 jeff cjs 39778950 Oct 21  2023 bronners_11.arw
-rw-r--r-- 1 jeff cjs 20399288 Jun 30 22:03 bronners_11.jpg
-rw-r--r-- 1 jeff cjs 36606358 Oct 21  2023 bronners_12.arw
-rw-r--r-- 1 jeff cjs 10895436 Jun 30 22:09 bronners_12.jpg
-rw-r--r-- 1 jeff cjs 38439916 Oct 21  2023 bronners_13.arw
-rw-r--r-- 1 jeff cjs 14554244 Jul 13 20:01 bronners_13.jpg
-rw-r--r-- 1 jeff cjs 37922748 Oct 21  2023 bronners_14.arw
-rw-r--r-- 1 jeff cjs 12683282 Jul 13 20:05 bronners_14.jpg
-rw-r--r-- 1 jeff cjs 37094742 Oct 21  2023 bronners_15.arw
-rw-r--r-- 1 jeff cjs 10315937 Jul 14 11:12 bronners_15.jpg
-rw-r--r-- 1 jeff cjs 37883250 Oct 21  2023 bronners_16.arw
-rw-r--r-- 1 jeff cjs  9174955 Jul 14 16:22 bronners_16.jpg
-rw-r--r-- 1 jeff cjs 38585528 Oct 21  2023 bronners_17.arw
-rw-r--r-- 1 jeff cjs 11760884 Jul 14 16:25 bronners_17.jpg
-rw-r--r-- 1 jeff cjs 38641462 Oct 21  2023 bronners_18.arw
-rw-r--r-- 1 jeff cjs 11869423 Jul 14 16:30 bronners_18.jpg
-rw-r--r-- 1 jeff cjs 38829932 Oct 21  2023 bronners_19.arw
-rw-r--r-- 1 jeff cjs 12573830 Jul 14 16:35 bronners_19.jpg
-rw-r--r-- 1 jeff cjs 39260562 Oct 21  2023 bronners_20.arw
-rw-r--r-- 1 jeff cjs 13548118 May 23 12:02 bronners_20.jpg
-rw-r--r-- 1 jeff cjs 38876960 Oct 21  2023 bronners_21.arw
-rw-r--r-- 1 jeff cjs 12872378 May 23 12:02 bronners_21.jpg
-rw-r--r-- 1 jeff cjs 38739862 Oct 21  2023 bronners_22.arw
-rw-r--r-- 1 jeff cjs 12563042 May 23 12:02 bronners_22.jpg
-rw-r--r-- 1 jeff cjs 38547174 Oct 21  2023 bronners_23.arw
-rw-r--r-- 1 jeff cjs 12274102 May 23 12:02 bronners_23.jpg
-rw-r--r-- 1 jeff cjs 38635340 Oct 21  2023 bronners_24.arw
-rw-r--r-- 1 jeff cjs 12532377 May 23 12:02 bronners_24.jpg
-rw-r--r-- 1 jeff cjs 38757902 Oct 21  2023 bronners_25.arw
-rw-r--r-- 1 jeff cjs 12724199 May 23 12:02 bronners_25.jpg
-rw-r--r-- 1 jeff cjs 38441502 Oct 21  2023 bronners_26.arw
-rw-r--r-- 1 jeff cjs 12113650 May 23 12:02 bronners_26.jpg

![Screenshot_2024-07-14_21-02-20](https://github.com/user-attachments/assets/4fdacf7e-5722-4e9f-af82-a5787e8972cb)

**Comment by xsdg** [comment]
Thanks for the extra info, and especially the screenshot!  I'd be really curious if this continues to manifest with thumbnails disabled, or in "Images as List" mode rather than "Images as Icons".

Basically, it's obvious that there's a bug here, but given that there aren't other folks reporting this same issue, it feels like there must be something that's causing it to show up for you but not for most other people.

Do your pictures tend to have a GPS location embedded?

**Comment by xsdg** [comment]
It's also really peculiar that Geeqie is ending up in the DDS loader at all.  The only codepath that leads to it is here: \
https://github.com/BestImageViewer/geeqie/blob/eedf9f9db467b9eb3ce1ae90dc8ec9c30e252f7a/src/image-load.cc#L754

And some sample JPEGs and ARWs I have on my system don't start with the three magic bytes to actually trigger that loader
```
$hd /.../nx7_7876.arw | head -2
00000000  49 49 2a 00 08 00 00 00  12 00 fe 00 04 00 01 00  |II*.............|
00000010  00 00 01 00 00 00 03 01  03 00 01 00 00 00 06 00  |................|

$hd /.../xt5_6666.jpg  | head -2
00000000  ff d8 ff e1 ff bc 45 78  69 66 00 00 49 49 2a 00  |......Exif..II*.|
00000010  08 00 00 00 0d 00 0f 01  02 00 09 00 00 00 aa 00  |................|
```

So Colin's hypothesis about a refcounting bug seems more likely than anything to do with the partially-written contents of the files.

**Comment by xsdg** [comment]
One last debugging step: if you can replicate this while running Geeqie with `--debug`, that should give some indication of where the train might have started to go off the tracks.

**Comment by caclark** [comment]
> /usr/lbin/Geeqie-240624/AppRun

When using an AppImage, you should execute the `.../squashfs-root/Apprun` file.
I get:
`file ~/bin/Geeqie-latest-x86_64-AppImage/squashfs-root/AppRun`
`/home/cclark/bin/Geeqie-latest-x86_64-AppImage/squashfs-root/AppRun: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=86bc6644a3eca0daef30e3b0ec672a7eb121044f, not stripped`

And:
`file ~/bin/Geeqie-latest-x86_64-AppImage/squashfs-root/usr/local/bin/geeqie`
`/home/cclark/bin/Geeqie-latest-x86_64-AppImage/squashfs-root/usr/local/bin/geeqie: ELF 64-bit LSB pie executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter lib64/ld-linux-x86-64.so.2, BuildID[sha1]=e3e02a82c5b7d1f7620cacdc108673c3e42f0d24, for GNU/Linux 3.2.0, with debug_info, not stripped`

Trying to execute the `.../squashfs-root/usr/local/bin/geeqie` file gives `cannot execute: required file not found`


For the last few issues of the AppImage, a stack-dump is automatically generated if you run the AppImage from the command line and there is a crash.

Would you please execute this (as yet un-registered) bug from a cold start to see if the stack-dumping is working OK?:
`File/New Collection`
Right-click on the Collection window and select `Save Collection As`
In `Save Collection As` insert `/tmp/xxxxx.txt`
Click `Save`

A stack-dump should be produced.

**Comment by jefferysmall** [comment]
xsdg:

I just tried this with the icon pane in "Images as List" mode and yes, it still crashed.
My pictures do not contain GPS data.

**Comment by jefferysmall** [comment]
xsdg:

Here is the output from:

**/usr/lbin/geeqie --debug --geometry=1852x2039+989+41 --disable-clutter  .**

This coredump resulted from repeatedly saving bronner_20.jpg.

[z.txt](https://github.com/user-attachments/files/16241114/z.txt)

**Comment by jefferysmall** [comment]
caclark:

I renamed **/usr/lbin/squashfs-root/** as **/usr/lbin/Geeqie-240624/** so that it doesn't get overridden by the next AppImage extraction.

**/usr/lbin/geeqie** is just a symlink to **/usr/lbin/squashfs-root/AppRun**

So I am executing the proper program.  I was previously confused, not realizing that the actual executable at  **./usr/local/bin/geeqie** was being **exec**ed, and I initially pointed gdb at the wrong source file.  The coredump does in fact work as I show above.  My second point of confusion was that I had long ago set **ulimit -c** to **0** by default.  Once I reset it to **unlimited** in the shell from which I was launching geeqie, I got the expected core dump.  So user error on this one, not a problem with the code.  Sorry about that.

**Comment by xsdg** [comment]
What do you get for `ls -l /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png`?

Also, could you repeat this with `--debug=2`?  It looks like there's some extra debug logging that exists specifically for thumbnail management.

In particular, the following pattern looks kind of peculiar, where Geeqie fails to load the thumbnail, tries to recreate it, and says that it succeeded, only to fail again later on.  (Note that there are two threads, so messages are interleaved a bit):
```
thumb progress: 9 of 26
new image loader 0x5608ba55a1c0, bufsize=4096 idle_loop=1
exif read /u/jeff/.cache/thumbnails/large/77e6f99d1fd7a32ed8c9a8bba2711f66.png, sidecar: -
Thread pool num threads: 2
freeing image loader 0x5608ba8b76a0 bytes_read=105247
thumb image done: /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_19.jpg
            from: /u/jeff/.cache/thumbnails/large/77e6f99d1fd7a32ed8c9a8bba2711f66.png
thumb progress: 10 of 26
new image loader 0x5608ba55a0f0, bufsize=4096 idle_loop=1
exif read /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png, sidecar: -
freeing image loader 0x5608ba55a0f0 bytes_read=0
thumb broken, unlinking: /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png
new image loader 0x5608ba55a020, bufsize=4096 idle_loop=1
exif read /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_20.jpg, sidecar: -
Found embedded icc profile in jpeg
Thread pool num threads: 2
freeing image loader 0x5608ba55a1c0 bytes_read=105251
Using custom jpeg loader
thumb image done: /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_20.jpg
            from: /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_20.jpg
thumb saving: /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_20.jpg
       saved: /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png
thumb progress: 11 of 26
new image loader 0x5608ba8b7500, bufsize=4096 idle_loop=1
exif read /u/jeff/.cache/thumbnails/large/731890edbeee1cd55bb1703891d6a49b.png, sidecar: -
Thread pool num threads: 1
freeing image loader 0x5608ba55a020 bytes_read=12276803
thumb image done: /u/jeff/Pictures/alpha/Michigan/Bronners/bronners_21.jpg
            from: /u/jeff/.cache/thumbnails/large/731890edbeee1cd55bb1703891d6a49b.png
thumb progress: 12 of 26
…
```

Then farther down, right before the crash, the same thing happens again (and including the stack trace for @caclark; it's very similar to the one from the core dump above):
```
new image loader 0x5608ba8b76a0, bufsize=4096 idle_loop=1
exif read /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png, sidecar: -
freeing image loader 0x5608ba8b76a0 bytes_read=0
thumb broken, unlinking: /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png
new image loader 0x5608ba55a0f0, bufsize=4096 idle_loop=1
Thread pool num threads: 2
   16.801792 (+00001.271583) pixbuf renderer updated - started drawing 0x5608ba564180, img: 4672x7008
   16.801824 (+00000.000032) pixbuf renderer updated - started drawing 0x5608ba564180, img: 4672x7008
   16.965300 (+00000.163476) image done
Using custom jpeg loader
freeing image loader 0x5608ba8b75d0 bytes_read=12276803
   17.137141 (+00000.171841) image load completed "/u/jeff/Pictures/alpha/Michigan/Bronners/bronners_20.jpg" (current)
   17.137193 (+00000.000052) pixbuf renderer done 0x5608ba564180
Stack trace (most recent call last) in thread 879166:
#10   Object "/usr/local/bin/Geeqie-240624/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7fcbbe66684f, in 
#9    Object "/usr/local/bin/Geeqie-240624/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7fcbbe5d4ac2, in 
#8    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7fcbbfb28ab0, in 
#7    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7fcbbfb2b713, in 
#6    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load-dds.cc:281", in <unknown>::ddsDecodeDXT3(uint width, uint height, unsigned char *buffer) [0x5608b787df2f]
#5    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load.cc:1393", in image_loader_get_error(ImageLoader *il) [0x5608b787d6a3]
#4    Source "/home/runner/work/geeqie/geeqie/build/../src/bar-gps.cc:778", in bar_pane_gps_view_state_changed_cb(ChamplainView *view, GParamSpec *, *gpointer data) [0x5608b7913f21]
#3    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7fcbbf1b62c5, in jpeg_start_decompress
#2    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7fcbbf1ae075, in 
#1    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7fcbbf1bbd70, in 
#0    Object "/usr/local/bin/Geeqie-240624/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7fcbbf1bbbcf, in jpeg_fill_bit_buffer
Bus error (Nonexisting physical address [0x7fcb9df79112])
Bus error(coredump)
```

**Comment by jefferysmall** [comment]
Here is output from:

/usr/lbin/geeqie --debug=2 --geometry=1852x2039+989+41 --disable-clutter  . > geeqee_debug_02.txt 2>&1

[geeqee_debug_02.txt](https://github.com/user-attachments/files/16242591/geeqee_debug_02.txt)

**Comment by jefferysmall** [comment]
101-> ls -l /u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png
ls: cannot access '/u/jeff/.cache/thumbnails/large/b51625f865de914d91a62cf2c151e4e4.png': No such file or directory

**Comment by xsdg** [comment]
Thanks, that's super interesting!  Unfortunately, it also looks like the debug write buffer isn't getting fully flushed before geeqie crashes, so the log is missing a bit of what's happening right before the crash.

Could you repeat the `--debug=2` in a directory containing:
1) `bronners_20.jpg` and `bronners_20.arw`
2) Separately, with just `bronners_20.jpg`

Basically, keeping it to one image pair will limit how much extraneous debug info is printed.  And independently, just the jpg because I'm curious if the sidecar mechanism (which associates raw files with their respective jpgs) might be related.

**Comment by jefferysmall** [comment]
Here is the debug output in a directory with two files: bronners_20.{arw,jpg}

[geeqie_debug_03.txt](https://github.com/user-attachments/files/16254803/geeqie_debug_03.txt)

**Comment by jefferysmall** [comment]
Here is the debug output in a directory with one file: bronners_20.jpg


[geeqie_debug_04.txt](https://github.com/user-attachments/files/16254813/geeqie_debug_04.txt)

**Comment by xsdg** [comment]
Could you repeat the single-file run with thumbnails disabled?  (Right-click on the file list and un-check "Show thumbnails")

Also, just so I can better understand what I'm seeing, how many saves from gimp is it usually taking to trigger the crash in these runs?  (Knowing if it's the first save versus if it took two or more would be the most helpful, even though an exact count would be extra handy)

**Comment by jefferysmall** [comment]
The only way to see the **Show trumbnails** option is to first select **Images as List**.  This is what I tried previously and it crashed, but I'll repeat it here with the single JPG file.  Here is what it looked like.

![Screenshot_2024-07-16_11-58-34](https://github.com/user-attachments/assets/54bfc2de-1eab-45f6-a61f-d4edff8c85e4)

Usually geeqie crashes after the second export, but sometimes it takes more.  It is also more likely to happen of the exports are issues rapidly, although that is not always necessary.  In this case I exported once.  Waited until it had completed and geeqie had updated.  Then exported a second time and geeqie crashed.  Here is the output.

[geeqie_debug_05.txt](https://github.com/user-attachments/files/16255411/geeqie_debug_05.txt)

**Comment by caclark** [comment]
For information only - the latest (maybe since May) AppImage also appends the stack trace to `/tmp/geeqie-crash.log`.

**Comment by jefferysmall** [comment]
And thanks for the note about /tmp/geeqie-crash.log.  That's very helpful.

**Comment by caclark** [comment]
> Speaking of AppImage, is there any way to be automatically notified when a new AppImage is generated and made available from the Homepage?

For quite a while there has been code that ran on start-up of Geeqie and shows a fade-out message that a new version is available.
It did not work.
I fixed the problem - and now it shows on every start-up, new version or not.
I will fix it again ... sometime.

[The fade-out message box should appear towards the upper-right side of the screen]

**Comment by jefferysmall** [comment]
here is the preferences menu from the 06-24-2024 AppImage.  I'm afraid I don;t see this option (am I blind??) and I have never seen the fade-out message on this or previous version.

![Screenshot_2024-07-18_11-34-40](https://github.com/user-attachments/assets/8be8d40b-bbd2-4de6-8a44-e975f8aa498c)

**Comment by caclark** [comment]
An AppImage can be run in two ways: directly from the AppImage or from the extracted executable.
To determine which is being used it checks:
In the first case it checks (in an indirect way) that `/tmp/.mount_Geeqie<unique string>` exists.
If not, it looks for `squashfs-root` in the executable path, which means it is an extracted AppImage.

If neither it is a normal executable.

**Comment by jefferysmall** [comment]
Not to confuse the party, but I just downloaded the latest AppImage file (Geeqie 2.4+git20240716-428bd7b1 GTK3), extracted it and tried to run it.

40-> squashfs-root/AppRun --debug=2 -geometry=1852x2039+989+41 --disable-clutter  .

It coredumps immediately with this output:
---------------------------------------------------------------------------------------------
../src/main.cc:663:parse_command_line_for_debug_option:debugging output enabled (level 2)
../src/main.cc:1287:main:    0.000631 (+00000.000631) main: gtk_init
Gtk-Message: 12:10:24.303: Failed to load module "xapp-gtk3-module"
../src/main.cc:1317:main:    0.009019 (+00000.008388) main: pixbuf_inline_register_stock_icons
../src/main.cc:1321:main:    0.173967 (+00000.164948) main: setting default options before commandline handling
../src/ui-fileops.cc:195:homedir:Home directory: /u/jeff
../src/ui-fileops.cc:195:homedir:Home directory: /u/jeff
../src/ui-fileops.cc:195:homedir:Home directory: /u/jeff
../src/main.cc:1332:main:    0.174200 (+00000.000233) main: mkdir_if_not_exists
../src/ui-fileops.cc:195:homedir:Home directory: /u/jeff
../src/main.cc:1353:main:    0.174376 (+00000.000176) main: parse_command_line

(geeqie:41999): Gdk-ERROR **: 12:10:25.181: The program 'geeqie' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadImplementation (server does not implement operation)'.
  (Details: serial 1507 error_code 17 request_code 20 (core protocol) minor_code 0)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Stack trace (most recent call last) in thread 42025:
#12   Object "/x/u/jeff/lib/Downloads/squashfs-root/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7ff5397ab84f, in 
#11   Object "/x/u/jeff/lib/Downloads/squashfs-root/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7ff539719ac2, in 
#10   Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac51ab0, in 
#9    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac54713, in 
#8    Source "/home/runner/work/geeqie/geeqie/build/../src/ui-utildlg.cc:497", in generic_dialog_new(const gchar *title, const gchar *role, GtkWidget *parent, gboolean auto_close, void *cancel_cb, *gpointer data) [0x562d2fa1ca26]
#7    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b149a26, in 
#6    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b149720, in 
#5    Object "/lib/x86_64-linux-gnu/libX11.so.6", at 0x7ff539522606, in XGetWindowProperty
#4    Object "/lib/x86_64-linux-gnu/libX11.so.6", at 0x7ff5395419f3, in _XError
#3    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b146eb3, in 
#2    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac29eb2, in g_log_structured_standard
#1    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac29cb2, in g_log_structured_array
#0    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac2e167, in g_log_writer_default
Trace/breakpoint trap (Signal sent by the kernel 0 0)
Trace/BPT trap(coredump)
---------------------------------------------------------------------------------------------

And the output in /tmp/geeqie-crash.log:
---------------------------------------------------------------------------------------------
Stack trace (most recent call last) in thread 42025:
#12   Object "/x/u/jeff/lib/Downloads/squashfs-root/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7ff5397ab84f, in 
#11   Object "/x/u/jeff/lib/Downloads/squashfs-root/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7ff539719ac2, in 
#10   Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac51ab0, in 
#9    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac54713, in 
#8    Source "/home/runner/work/geeqie/geeqie/build/../src/ui-utildlg.cc:497", in generic_dialog_new(const gchar *title, const gchar *role, GtkWidget *parent, gboolean auto_close, void *cancel_cb, *gpointer data) [0x562d2fa1ca26]
#7    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b149a26, in 
#6    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b149720, in 
#5    Object "/lib/x86_64-linux-gnu/libX11.so.6", at 0x7ff539522606, in XGetWindowProperty
#4    Object "/lib/x86_64-linux-gnu/libX11.so.6", at 0x7ff5395419f3, in _XError
#3    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libgdk-3.so.0", at 0x7ff53b146eb3, in 
#2    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac29eb2, in g_log_structured_standard
#1    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac29cb2, in g_log_structured_array
#0    Object "/x/u/jeff/lib/Downloads/squashfs-root/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7ff53ac2e167, in g_log_writer_default
---------------------------------------------------------------------------------------------

**Comment by jefferysmall** [comment]
I just downloaded the latest App Image (Geeqie 2.4+git20240716-428bd7b1 GTK3) and tried it.  Not unsurprisingly, it coredumps just like the earlier version.  This was in the directory with the single image file.  In this case it coredumped on the third export from GIMP.

Here is the content of /tmp/geeqie-crash.log:
---------------------------------------------------------------------------
Stack trace (most recent call last) in thread 79482:
#10   Object "/usr/local/bin/Geeqie-240716/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7f083b99484f, in 
#9    Object "/usr/local/bin/Geeqie-240716/runtime/compat/lib/x86_64-linux-gnu/libc.so.6", at 0x7f083b902ac2, in 
#8    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7f083ce51ab0, in 
#7    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libglib-2.0.so.0", at 0x7f083ce54713, in 
#6    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load-dds.cc:281", in <unknown>::ddsDecodeDXT3(uint width, uint height, unsigned char *buffer) [0x555d14043bff]
#5    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load.cc:1405", in image_loader_get_error(ImageLoader *il) [0x555d14043373]
#4    Source "/home/runner/work/geeqie/geeqie/build/../src/bar-gps.cc:771", in bar_pane_gps_view_state_changed_cb(ChamplainView *view, GParamSpec *, *gpointer data) [0x555d140dc1a1]
#3    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7f083c4e42c5, in jpeg_start_decompress
#2    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7f083c4dc075, in 
#1    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7f083c4ec08e, in 
#0    Object "/usr/local/bin/Geeqie-240716/usr/lib/x86_64-linux-gnu/libjpeg.so.8", at 0x7f083c4e9bcf, in jpeg_fill_bit_buffer
---------------------------------------------------------------------------

And the debug output:
---------------------------------------------------------------------------
[geeqie_debug_07.txt](https://github.com/user-attachments/files/16289062/geeqie_debug_07.txt)
---------------------------------------------------------------------------

**Comment by jefferysmall** [comment]
Every so often I download and try a newer AppImage version (extracted).  I just downloaded **2.4+git20240902-a3abbc65** and it is still crashing when the thumbnail of the actively edited image (in GIMP) is exported.

I'm willing to do more testing if anyone has further suggestion of what info to collect.  I'd really like to see this issue get resolved.

Regards.

**Comment by jefferysmall** [comment]
I'm back to editing a lot of new pictures in GIMP and I am often still having this problem of qeeqie exiting when GIMP exports an edited jpg file, overwriting the existing one.

In the latest AppImage version, I'm no longer seeing geeqie exit when I use a utility to rapidly renumber files (e.g., dsc_01.jpg, dec_12.jpg,  dsc_13.jpg, dsc_20.jpg, etc), including when the filenames overlap.  When overlapping, this tool moves files to temporary filenames and then renames those files with the appropriate numbers.  This works while geeqie is viewing the directory, but with overlapping names, it can still fail to properly update the thumbnail, leaving the old thumbnail in place.  I can now do a "touch *" in the directory to force the thumbnails to update and I never see geeqie crash when this is done.

So there is something special happening when GIMP exports a JPG file that causes geeqie to frequently, but not always, crash.  So I'm following up with this latest note to suggest that this may need to be tested against GIMP in order to better understand the steps that are occurring beneath the hood that is triggering the geeqie problem.  I can see that early in the export process the file is being removed, or possible zeroed out (I'd look at that), and geeqie displays a broken image icon as the new export is being written out to disk.  It is somewhere during or at the end of the export operation when geeqie crashes.

Happy New Year to everyone.

**Comment by jefferysmall** [comment]
Follow up:  I'm now trying out the AppImage:
Geeqie 2.5+git20250103-f41bec46 GTK3

My first impression was that this version is much more stable than the previous 2.4 AppImage versions, but after using it for a few days with GIMP, I have to unfortunately report that it continues to crash when exporting images as reported above -- not on every export, but still frequently.  Fixing this is at the top of my request list.

**Comment by caclark** [comment]
@jefferysmall 
The good news is that I have found a way to reproduce the fault.
The bad news is that it is a problem with a background function and a thread function accessing the same data.
It will take a while but at least there is a faint light at the end of the tunnel.

**Comment by jefferysmall** [comment]
Colin:

Great news!  I'll take that faint light.  :-)

As always, thanks for your relentless pursuit of these nasty bugs.

**Comment by jefferysmall** [comment]
Colin: I just downloaded Geeqie 2.5+git20250109-97d189c3 GTK3 and was trying it out.  It started out appearing to be a bit more stable but then it crashed while exporting a GIMP file.  However, the geeqie window remained on the screen and now there was a pop-up window that said:

--------------------------------------------------------------------
Sorry - Geeqie has crashed
Crash log file: /tmp/geeqie-crash.log
--------------------------------------------------------------------

However, I could not dismiss or close this pop-up.  Geeqie was stuck and I had to manually terminate the application.

Here is the content of /tmp/geeqie-crash.log:

--------------------------------------------------------------------
Stack trace (most recent call last) in thread 162458:
#10   Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x789dd7129c3b, in 
#9    Object "/lib/x86_64-linux-gnu/libc.so.6", at 0x789dd709ca93, in 
#8    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libglib-2.0.so.0", at 0x789dd8a1fc81, in 
#7    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libglib-2.0.so.0", at 0x789dd8a25541, in 
#6    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load.cc:1190", in image_loader_thread_run(*gpointer data, *gpointer) [0x64f27ac75ac0]
#5    Source "/usr/include/c++/13/bits/unique_ptr.h:199", in  [0x64f27ac75223]
        196:       _GLIBCXX23_CONSTEXPR
        197:       pointer&   _M_ptr() noexcept { return std::get<0>(_M_t); }
        198:       _GLIBCXX23_CONSTEXPR
      > 199:       pointer    _M_ptr() const noexcept { return std::get<0>(_M_t); }
        200:       _GLIBCXX23_CONSTEXPR
        201:       _Dp&       _M_deleter() noexcept { return std::get<1>(_M_t); }
        202:       _GLIBCXX23_CONSTEXPR
#4    Source "/home/runner/work/geeqie/geeqie/build/../src/image-load-jpeg.cc:362", in gboolean write(const guchar *buf, gsize &chunk_size, gsize count, GError **error) [0x64f27ad12e51]
#3    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libjpeg.so.8", at 0x789dd7d5eb50, in jpeg_start_decompress
#2    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libjpeg.so.8", at 0x789dd7d5655e, in 
#1    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libjpeg.so.8", at 0x789dd7d666a9, in 
#0    Object "/usr/local/bin/Geeqie-250109/usr/bin/../lib/libjpeg.so.8", at 0x789dd7d64551, in jpeg_fill_bit_buffer
--------------------------------------------------------------------

Hope this helps.

**Comment by caclark** [comment]
On Saturday, 11 January 2025 at 21:05:54 -00:00, cjsmall ***@***.***> wrote:

> 
> Colin: I just downloaded Geeqie 2.5+git20250109-97d189c3 GTK3 and was trying it out. It started out appearing to be a bit more stable but then it crashed while exporting a GIMP file. However, the geeqie window remained on the screen and now there was a pop-up window that said:
> 
> 
> Sorry - Geeqie has crashed
> Crash log file: /tmp/geeqie-crash.log
> However, I could not dismiss or close this pop-up. Geeqie was stuck and I had to manually terminate the application.
> Here is the content of /tmp/geeqie-crash.log:
> 
The dialog is just to inform users that there is a crash log in /tmp/

The "Close" button in the dialog should terminate Geeqie.

**Comment by jefferysmall** [comment]
The close button isn't working for me and I've tried on multiple occasions.  My assumption was that since geeqie crashed, after throwing up the dialog there was no controlling process left to take down wither the pop0up or the main window.  After clicking on the [X] (close) icon numerous times, my window manager says something like: This process is busy and not responding.  Do you want to close the window?

I've found some other odd behavior between multiple instances started with --new-window which I want to test more before reporting to you.  I'll do that on the developer mailing list where that is being discussed.

I may switch over to launching each instance of geeqie as its own process and see how that works.

**Comment by caclark** [comment]
Running Geeqie on a different computer results in the error message dialog not closing.
A workaround (for me) is to press the escape key.

**Comment by xsdg** [comment]
> [@jefferysmall](https://github.com/jefferysmall) The good news is that I have found a way to reproduce the fault. The bad news is that it is a problem with a background function and a thread function accessing the same data. It will take a while but at least there is a faint light at the end of the tunnel.

Colin: just for curiosity, what's the shared data and where is it being concurrently accessed from?

**Comment by caclark** [comment]
I am using a jpeg as a test image. The seg. fault is consistently on line 362 of `image-load-jpeg.cc`
`	jpeg_start_decompress(&cinfo);`

If  I comment out lines 148 to 164 of `file_cache_remove_fd()` in `filecache.cc` there is no fault.

I do not understand where the problem is. I was thinking just to use a mutex to prevent the two threads accessing whatever the data is - but that is a bit of a primitive solution  [my style of programming :-)  ]

**Comment by jefferysmall** [comment]
I'm now using:

Geeqie 2.5+git20250402-22ea4833 GTK3

and the crash problem still exists.  I just wanted to give feedback as I keep moving forward.  I appreciate your working on this nasty bug.

**Comment by xsdg** [comment]
This should be fixed with commits from May 17th.  Thanks for reporting this way back when so that it was easier to spot new clues when they arose.

(Also, please re-open if you see this issue again after those commits)

**Comment by jefferysmall** [comment]
Well I was really excited to see this so I downloaded the latest build (2.7+git20260524-ca73f9b7), loaded a jpg image into GIMP 3.2.4, and proceeded to export the image.  Geeqie immediately crashed.  I reloaded the image and tried re-exporting it.  It worked a couple of times and then geeqie crashed again!  I can repeat this over and over.  This is the same behavior I've been seeing.  Sorry to have to report this.  Please let me know if I can provide any additional information.

**Comment by jefferysmall** [comment]
I don't have permission to reopen the issue.

**Comment by xsdg** [comment]
> I just downloaded the latest App Image (Geeqie 2.4+git20240716-428bd7b1 GTK3) and tried it. This is the extracted version. Not unsurprisingly, it coredumps just like the earlier version. This was in the directory with the single image file. In this case it coredumped on the third export from GIMP.
> 
> […]
> 
> And the debug output:
> 
> [geeqie_debug_07.txt](https://github.com/user-attachments/files/16289062/geeqie_debug_07.txt)

@jefferysmall If you're able to get another debug trace like this one, that would be really helpful.  Basically, the debug messages look like they're triggering the bug codepaths that I fixed, so I'm curious if there's another codepath that triggers the same bug, or whether it's a different bug.

That said, given that you can reproduce this so reliably, I'm also curious if you can get it to happen with any other file format, or if it's specifically only JPEG.  (In particular, I'm wondering whether this might actually be a bug in some jpeg-specific geeqie code or in libjpeg itself)

**Comment by jefferysmall** [comment]
I took the **jpg** image I was testing with yesterday, and converted it to a **tif**.  I loaded the tif into GIMP and repeatedly saved it while viewing the **tif** image in geeqie.  While much more stable than the **jpg**, I repeatedly exported the **tif** file and I am able to repeatedly get geeqie to crash.

One thing I notice is that GIMP exports the **tif** file very quickly but takes a lot more time to export the **jpg**, even though the **tif** file is over twice the size of the jpg.  Because of the intermittent nature of the crashing, I have always suspected that there is a timing component to the crash, where GIMP is deleting or zeroing out the file during the export, and geeqie is "triggered" at some point during this, or the subsequent replacement of the contents.  (I have not looked at GIMP code to see what it actually does.)  Since GIMP writes the **tif** file so much more quickly, I have strong suspicions that this is why it "fools" geeqie and triggers the crash much less frequently -- but it's only a theory.  And as Galileo would say: And yet it crashes.  :-)

I'll try to run geeqie and examine the core dump, but I need to change my environment to do that as I have coredumps blocked by default.  When I get new results I'll post them here using the jpg image as that will be easier to trigger with less debugging output.  More later.

**Comment by jefferysmall** [comment]
I set up my environment to allow core dumps.  However, when I run geeqie and force it to crash with my exporting the GIMP image, I get no core file in either **/var/crash** or **/var/lib/apport/coredump**.

Based upon the notes above I tried running:

% /usr/lbin/Geeqie-260524/AppRun -Ddevel=enabled --geometry=1852x2039+989+41   .
and am again getting:  "Unknown option -Ddevel=enabled"

What am I missing?  What do I need to do to generate the crash file?

**Comment by xsdg** [comment]
I was finally able to repro this.  Here are the steps I used.  I don't know which ones might be important:

1) Create a large file (8000x8000) in gimp and save as a JPG to a test directory.
2) Open that directory and file in Geeqie
3) Make sure that the Geeqie window is visible (in my case, on the same workspace as GIMP, and I used "Always On Top" mode in my window manager)
4) Re-export the file from GIMP (using Ctrl+E)
5) Eventually, Geeqie crashes.

**Comment by xsdg** [comment]
Here's the backtrace with libjpeg62-turbo-dbgsym installed:
```
Thread 11 "pool-4" received signal SIGBUS, Bus error.
[Switching to Thread 0x7fffed29e6c0 (LWP 743261)]
jpeg_fill_bit_buffer (state=state@entry=0x7fffed29d140, get_buffer=17434341793721197455, bits_left=6, nbits=nbits@entry=0) at ./src/jdhuff.c:312
⚠️ warning: 312	./src/jdhuff.c: No such file or directory
(gdb) bt
#0  jpeg_fill_bit_buffer (state=state@entry=0x7fffed29d140, get_buffer=17434341793721197455, bits_left=6, nbits=nbits@entry=0) at ./src/jdhuff.c:312
#1  0x00007ffff70b5eaf in decode_mcu_AC_refine (cinfo=<optimized out>, MCU_data=<optimized out>) at ./src/jdphuff.c:549
#2  0x00007ffff709a4c7 in consume_data (cinfo=0x7fffed29d7d0) at ./src/wrapper/../jdcoefct.c:236
#3  0x00007ffff7095f81 in jpeg_start_decompress (cinfo=0x7fffed29d7d0) at ./src/wrapper/../jdapistd.c:74
#4  0x00005555556aaf5a in ImageLoaderJpeg::write
    (this=0x7fffd80040e0, buf=0x7fffcbd9b000 "\377\330\377", <incomplete sequence \340>, chunk_size=@0x7fffed29da78: 4096, count=2506770, error=<optimized out>)
    at ../src/image-load-jpeg.cc:359
#5  0x000055555562905d in image_loader_begin (il=il@entry=0x5555588ee470) at ../src/image-load.cc:858
#6  0x00005555556292d4 in image_loader_thread_run (data=0x5555588ee470) at ../src/image-load.cc:1151
#7  0x00007ffff720de2a in ??? () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#8  0x00007ffff720d676 in ??? () at /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#9  0x00007ffff629dda9 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:448
#10 0x00007ffff631ce08 in __GI___clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78
(gdb) info threads
  Id   Target Id                                            Frame 
  1    Thread 0x7ffff568af80 (LWP 743208) "geeqie"          __syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
  2    Thread 0x7ffff4dff6c0 (LWP 743211) "[pango] fontcon" syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
  3    Thread 0x7fffeffff6c0 (LWP 743212) "pool-spawner"    syscall () at ../sysdeps/unix/sysv/linux/x86_64/syscall.S:38
  4    Thread 0x7fffef7fe6c0 (LWP 743213) "gmain"           __syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
  5    Thread 0x7fffeeffd6c0 (LWP 743214) "gdbus"           __syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
  9    Thread 0x7fffe61ff6c0 (LWP 743223) "dconf worker"    __syscall_cancel_arch () at ../sysdeps/unix/sysv/linux/x86_64/syscall_cancel.S:56
* 11   Thread 0x7fffed29e6c0 (LWP 743261) "pool-4"          jpeg_fill_bit_buffer (state=state@entry=0x7fffed29d140, get_buffer=17434341793721197455, bits_left=6, nbits=nbits@entry=0)
    at ./src/jdhuff.c:312
(gdb) 
```

And from the debian source package, here's what that function looks like:
```C
GLOBAL(boolean)
jpeg_fill_bit_buffer(bitread_working_state *state,
                     register bit_buf_type get_buffer, register int bits_left,
                     int nbits)
/* Load up the bit buffer to a depth of at least nbits */
{
  /* Copy heavily used state fields into locals (hopefully registers) */
  register const JOCTET *next_input_byte = state->next_input_byte;
  register size_t bytes_in_buffer = state->bytes_in_buffer;
  j_decompress_ptr cinfo = state->cinfo;

  /* Attempt to load at least MIN_GET_BITS bits into get_buffer. */
  /* (It is assumed that no request will be for more than that many bits.) */
  /* We fail to do so only if we hit a marker or are forced to suspend. */

  if (cinfo->unread_marker == 0) {      /* cannot advance past a marker */
    while (bits_left < MIN_GET_BITS) {
      register int c;

      /* Attempt to read a byte */
      if (bytes_in_buffer == 0) {
        if (!(*cinfo->src->fill_input_buffer) (cinfo))
          return FALSE;
        next_input_byte = cinfo->src->next_input_byte;
        bytes_in_buffer = cinfo->src->bytes_in_buffer;
      }
      bytes_in_buffer--;  /**** THIS IS LINE 312 ****/
      c = *next_input_byte++;
```

**Comment by xsdg** [comment]
The dereference on line 313 definitely looks a little suspicious.  That said, it also feels likely-relevant that this is running on a background image loader thread.

The internal Geeqie state looks pretty reasonable at first glance:
```
(gdb) frame 4
#4  0x00005555556aaf5a in ImageLoaderJpeg::write (this=0x7fffd80040e0, buf=0x7fffcbd9b000 "\377\330\377", <incomplete sequence \340>, chunk_size=@0x7fffed29da78: 4096, count=2506770, 
    error=<optimized out>) at ../src/image-load-jpeg.cc:359
359		jpeg_start_decompress(&cinfo);
(gdb) p cinfo
$2 = {err = 0x7fffed29d3c0, mem = 0x7fffd8005ec0, progress = 0x0, client_data = 0x7ffff7240d09, is_decompressor = 1, global_state = 203, src = 0x7fffd8004c00, image_width = 8000, 
  image_height = 8000, num_components = 3, jpeg_color_space = JCS_YCbCr, out_color_space = JCS_RGB, scale_num = 1, scale_denom = 1, output_gamma = 1, buffered_image = 0, raw_data_out = 0, 
  dct_method = JDCT_ISLOW, do_fancy_upsampling = 1, do_block_smoothing = 1, quantize_colors = 0, dither_mode = JDITHER_FS, two_pass_quantize = 1, desired_number_of_colors = 256, 
  enable_1pass_quant = 0, enable_external_quant = 0, enable_2pass_quant = 0, output_width = 8000, output_height = 8000, out_color_components = 3, output_components = 3, 
  rec_outbuf_height = 1, actual_number_of_colors = 0, colormap = 0x0, output_scanline = 0, input_scan_number = 8, input_iMCU_row = 277, output_scan_number = 0, output_iMCU_row = 0, 
  coef_bits = 0x7fffd8007980, quant_tbl_ptrs = {0x7fffd8004c40, 0x7fffd8004ce0, 0x0, 0x0}, dc_huff_tbl_ptrs = {0x7fffd8004d80, 0x7fffd8004ea0, 0x0, 0x0}, ac_huff_tbl_ptrs = {
    0x7fffd8004fc0, 0x7fffd8003f20, 0x0, 0x0}, data_precision = 8, comp_info = 0x7fffd8006cc0, progressive_mode = 1, arith_code = 0, arith_dc_L = '\000' <repeats 15 times>, 
  arith_dc_U = '\001' <repeats 16 times>, arith_ac_K = '\005' <repeats 16 times>, restart_interval = 0, saw_JFIF_marker = 1, JFIF_major_version = 1 '\001', JFIF_minor_version = 1 '\001', 
  density_unit = 1 '\001', X_density = 300, Y_density = 300, saw_Adobe_marker = 0, Adobe_transform = 0 '\000', CCIR601_sampling = 0, marker_list = 0x0, max_h_samp_factor = 1, 
  max_v_samp_factor = 1, min_DCT_scaled_size = 8, total_iMCU_rows = 1000, sample_range_limit = 0x7fffd8006ee0 "", comps_in_scan = 1, cur_comp_info = {0x7fffd8006d80, 0x0, 0x0, 0x0}, 
  MCUs_per_row = 1000, MCU_rows_in_scan = 1000, blocks_in_MCU = 1, MCU_membership = {0, 1, 2, 0, 0, 0, 0, 0, 0, 0}, Ss = 1, Se = 63, Ah = 1, Al = 0, unread_marker = 0, 
  master = 0x7fffd8004b60, main = 0x7fffd8008300, coef = 0x7fffd8007f80, post = 0x7fffd80074e0, inputctl = 0x7fffd8004b20, marker = 0x7fffd8004a00, entropy = 0x7fffd8007900, 
  idct = 0x7fffd8007520, upsample = 0x7fffd80073c0, cconvert = 0x7fffd8007360, cquantize = 0x0}
```

The `count` argument to `::write` matches the (fully-written) on-disk size of the jpeg:
```
$l test_gradient.jpg 
-rw-r--r-- 1 xsdg xsdg 2506770 May 27 01:00 test_gradient.jpg
```

**Comment by xsdg** [comment]
Running under ASAN makes the crash _much_ less reliable, but it still eventually happened.  ASAN said:
```
==746081==ERROR: AddressSanitizer: heap-use-after-free on address 0x7c6fdb310e80 at pc 0x55d237bf1c2c bp 0x7fff859c9d30 sp 0x7fff859c9d28
WRITE of size 8 at 0x7c6fdb310e80 thread T0
```

**Comment by jefferysmall** [comment]
I don't have any problem getting geeqie to crash.  What I'm not getting is a core file to examine and run the backtrace against.  I've changed my environment to allow unlimited core file sizes to be created, but I'm not finding anything.  Is there a special setting or argument required to generate the core file?  I'm happy to produce output here on my Xubuntu system if I can.

In the meantime, it's good that you are seeing the same thing I am.  Let me know what next step I can take in helping.

**Comment by xsdg** [comment]
Okay, so good news and bad news.  The good news is that (thanks in part to very generous assistance from the libjpeg-turbo maintainer) I think I understand what's happening.  The bad news is that I don't really know how to fix it yet

Here's another question about the exact same behavior we seem to be running into:
https://unix.stackexchange.com/questions/519850/mmap-effect-of-other-processes-writing-to-a-file-previously-mapped-read-only

And in particular, Geeqie does `mmap` files into memory:
https://github.com/BestImageViewer/geeqie/blob/master/src/ui-fileops.cc#L981

And the `mmap` documentation specifies the following about the `MAP_PRIVATE` mode (which Geeqie uses):
> It is  unspecified  whether changes made to the file after the mmap() call are visible in the mapped region.

One option would be to trap `SIGBUS` when an `ImageLoaderBackend` is running.  But that has the downside that we might silence crashes that really _should_ propagate and kill the process (for instance, due to memory corruption).  More research needed…

**Comment by xsdg** [comment]
Blog post about how SublimeText (a cross-platform text editor) dealt with this:
https://www.sublimetext.com/blog/articles/use-mmap-with-care

**Comment by xsdg** [comment]
https://stackoverflow.com/questions/41416028/why-does-a-read-operation-on-a-memory-mapped-zero-byte-file-lead-to-sigbus
cites the POSIX standard:
https://pubs.opengroup.org/onlinepubs/9699919799/functions/mmap.html

Which also notes:
> If the size of the mapped file changes after the call to mmap() as a result of some other operation on the mapped file, the effect of references to portions of the mapped region that correspond to added or removed portions of the file is unspecified.

And later…
> Memory access within the mapping but beyond the current end of the underlying objects may result in SIGBUS signals being sent to the process. The reason for this is that the size of the object can be manipulated by other processes and can change at any moment. The implementation should tell the application that a memory reference is outside the object where this can be detected; otherwise, written data may be lost and read data may not reflect actual data in the object.

**Comment by xsdg** [comment]
Some guidelines/best practices around signal handling on Linux
https://stackoverflow.com/questions/17942034/simple-linux-signal-handling

**Comment by jefferysmall** [comment]
xsdg:  That's some impressive investigation work you've done there.  I read these links in detail and am surprised that mmap has these sorts of problems that haven't been addressed internally in its own code.

I don't know the best path to take.  Could the SIGBUS signal be trapped and then perform some other test to see if the problem is related to the underlying file changing external to geeqie?  I suppose there is the possibility of a race condition as GIMP (or whatever program is modifying the program) takes time to fully replace the file.

And just FYI, as I have reported before, I have a program rapidly that renames files in a directory.  Usually this is a renumbering of files with <basename>_###.jpg where there are holes in the sequence or the starting number in the sequence is changed.  The program renames all the files with the desired numbers.  I have never seen geeqie crash as a result of this operation, so there is something different going on when this renaming happens, even if one of the renamed files is being displayed.  However, regarding the renumbering/renaming, if there is overlap in the pre and post numbering, geeqie can fail to update the thumbnails and sometimes I cannot get it to update in place.  I typically need to step geeqie out of the directory, touch all the files, and then step back into the directory and let all the thumbnails be rebuilt.  This is a separate issue, but wanted to mention the file renaming as NOT causing a crash.

**Comment by xsdg** [comment]
Okay, so it looks like part of what's going on here is a shortcoming of institutional memory 🙃

Code to deal with this situation was added to Geeqie a long time ago (2009) in https://github.com/BestImageViewer/geeqie/commit/96c8762cd141c1fca795e203a96307c80580d7f4 .  Even at the time of writing, Vladimir pointed out that the recovery strategy was a little questionable.

And then in 2022, it was disabled by @gdt in https://github.com/BestImageViewer/geeqie/commit/ffb0e76064eb444f4d7589dbe17bc36abece34a6 due to undefined behavior (which was likely more obvious after I had ported Geeqie to C++ earlier in 2022).  More context in #1052

And so that explains the timing of @jefferysmall finding and reporting this issue in 2023 (given the infrequency of Geeqie releases and such)

**Comment by xsdg** [comment]
I don't understand what the original SIGBUS handler was doing:
```c
static void sigbus_handler_cb(int signum, siginfo_t *info, void *context)
{
	unsigned long pagesize = sysconf(_SC_PAGE_SIZE);
	DEBUG_1("SIGBUS %p", info->si_addr);
	mmap((void *)(((unsigned long)info->si_addr / pagesize) * pagesize), pagesize, PROT_READ | PROT_WRITE, MAP_FIXED | MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
}
```

For comparison, at that same commit, this is the code the image loader used to originally mmap the file:
```c
il->mapped_file = mmap(0, il->bytes_total, PROT_READ|PROT_WRITE, MAP_PRIVATE, load_fd, 0);
```

I think what would make sense going forward is to set up a SIGBUS handler that allows Geeqie to somehow just treat the SIGBUS as a read error, which it already has plenty of ways to handle.  In particular, Geeqie can deal just fine with files that are deleted.

**Comment by xsdg** [comment]
Okay, from the `mmap` manpage:
>        MAP_ANONYMOUS
>               The mapping is not backed by any file; its contents are initial‐
>               ized to zero.  The fd argument is ignored; however, some  imple‐
>               mentations require fd to be -1 if MAP_ANONYMOUS (or MAP_ANON) is
>               specified, and portable applications should  ensure  this.   The
>               offset  argument  should  be zero.  Support for MAP_ANONYMOUS in
>               conjunction with MAP_SHARED was added in Linux 2.4.

So it looks like in response to `SIGBUS`, the original handler was re-mapping the faulted page to be empty, which would presumably continue for each successive page until the reader/parser either errored out due to a parse error, or finished reading all zeroes and returned.  That also doesn't seem like a great solution.  I still feel like it would be good to figure out some way for this to just act like a read error.

**Comment by xsdg** [comment]
Quoting from https://www.reddit.com/r/programming/comments/24fsqw/comment/ch7askv/
> If you want to handle errors, then you have to be prepared to receive SIGBUS. And probably the only reasonable thing you can do to recover from mmap-related SIGBUS is to longjmp() to a region of code where you're no longer trying to access the mapped region. In other words, it's a royal pain in the ass to write mmap code which is robust to errors.

Which makes sense.  Because the attempted failed read isn't actually from the code in question executing a syscall; it's from an attempted memory access (which triggers the syscall in the background).  And there's no error-handling mechanism other than a signal for a failed memory access.

That said, for Geeqie in particular, we use `mmap` in few enough places that this should be manageable.  Also, we _should_ be able to kill (and maybe reschedule) an ImageLoaderBackend thread and have Geeqie to continue to operate correctly.

But also, because the kernel can choose any random thread to handle a signal, maybe we should kill+reschedule all ImageLoaderBackend threads when a `SIGBUS` is received, and just make sure that the main thread isn't allowed to receive `SIGBUS`?  I feel like that would be a 90% solution that wouldn't affect the majority of users, and would slow Geeqie down but avoid a crash for the users who do need to handle truncations during image parsing

**Comment by xsdg** [comment]
Thankfully, it looks like my last paragraph in the last response (about the kernel choosing a thread at random) was mistaken*.  So this means that we'll receive the `SIGBUS` on the thread that faulted, which can then just `longjmp` to reschedule itself and terminate.

Then we can block `SIGBUS` on the main thread, which means that if we happen to encounter a `SIGBUS` on the main thread, it should just trigger the default behavior, which is to crash.
 
(\* Asked in the [linuxquestions](https://www.reddit.com/r/linuxquestions/comments/1tx7ne6/is_sigbus_from_an_mmap_fault_guaranteed_to_be_a/) and [c_programming](https://www.reddit.com/r/C_Programming/comments/1tx7ub1/is_sigbus_from_an_mmap_fault_guaranteed_to_be_a/) subreddits and the consensus is that `SIGBUS` will be delivered to the thread that triggered it, because (1) that's what the kernel code does on Linux, (2) that's apparently what other platforms do too, and (3) that's the only behavior that makes any sense anyway, particularly since the `SIGBUS` means that the faulting thread itself can't possibly make any forward progress)

**Comment by jefferysmall** [comment]
I was really excited that this issue appeared to be almost resolved, but there has been no further mentions here.  Just checking in.  Is this fix still moving forward?  Regards.

**Comment by xsdg** [comment]
Unfortunately, we paused all other work to focus on the fairly invasive [GTK4 migration](https://github.com/BestImageViewer/geeqie/discussions/2446) (and then I had some other stuff come up, so I've been out of commission for a bit on top of that)

**Comment by jefferysmall** [comment]
OkK thanks for the update.
